### ISSUE TITLE
Custom extractor

### ISSUE THREAD

**Issue by RaananHadar** [pachyderm/pachyderm#5359]
Note: this is a "future" level feature request. But I think its an important one.

**What is the goal / desired outcome?**
Improve performance and maintainability for scenarios that have to access data outside of Pachyderm.

**If there is a way to accomplish this today via workaround, what does that require?**
Pachyderm can access external data effectively but it can be a lot more efficient.

Lets describe the typical scenario which this issue deals with:
A pipeline downloads a lot of data that does not need to be version controlled and or exists outside of Pachyderm then transforms it on a GPU. Note that this issue is also relevant for non-GPU pipelines but I think that this is a common and important use case where this becomes very important: the GPU is very fast and this may become a performance bottleneck compared to alternative implementations.

The current way to implement this in Pachyderm:

* Implement a small "monolith" pipeline that runs in a single container that does both the downloading and the transform. This has 2 big issues:
     1. Utilization is low as time is wasted in IO. In our scenario, the GPU is very fast and this may become a performance bottleneck. Since it is common in k8s to exclusively bind the GPU to the user container, this leads to low utilization of an expensive resource.
     2. This container is a monolith that should be decoupled and becomes complex to maintain.
* Have a 2 stage Pachyderm pipeline: an extractor stage followed by a transformer stage. In this case, the software is more de-coupled but the data is needlessly persisted in Pachyderm and we pay for the extra IO to pull data into Pachyderm and read the data from pfs to the GPU node which is very inefficient. The severity of the persisted data will be reduced in the future when Pachyderm will have data retention BUT, it still doesn't solve the performance issue which is the bigger problem of the two.

**(Optional) What is your proposal for a feature to solve this?**
Pachyderm solves IO for data inside Pachyderm quite elegantly: by using a storage container as a sidecar and allowing data pre-fetching for the next datum to happen concurrently to user code run by using `max_queue_size` in the pipeline spec. My solution is to allow to compose user code to the storage sidecar: basically implement a user code that fetches the data for the next pipeline.

My roughly proposed solution will be to add an "extract" and "extract_input" fields to the pipeline spec. And have the extractor save the data in a specific directory such as "/pfs/in":
```
"extract": {
    "cmd": [ "python3", "/downloader.py" ],
    "image": "pachyderm/opencv_downloader"
},
"extract_input": {
    "pfs": {
      "repo": "images",
      "glob": "/*"
    }
}
```

**Comment by jdoliner** [comment]
Thanks for the feature request @RaananHadar, I think I understand the use case. I don't think I understand your proposal to fix it though. The "extract_input" would have to run on each worker before we could process any datums. During that time it would be monopolizing the GPU which is the thing we're trying to avoid. It would have an advantage over downloading the data for each datum because you only need to do it once per worker. But this is pretty similar to having your user code download the data to a /tmp directory if it doesn't fine it there. We don't clean up the filesystem between datum runs which allows you to cache data like this. But still this doesn't solve the fundamental problem that the time spent downloading the data will monopolize the GPU without actually using it.

One thing I have seen work well for this is loading the dataset onto a network attached storage device like EBS. Then you can mount the EBS volume into your container using the `pod_patch` and you won't have to waste time downloading it. If you mount it readonly you'll be able to mount it on multiple pods concurrently so you can still use parallelism.

**Comment by RaananHadar** [comment]
So let me expand on this. I am wondering if there is a way to extend the behaviour of storage container in Pachyderm.

Since a pod is scheduled on a single node in k8s, they can each have different resource requests. So while the worker container will ask for the bulk of the compute resources (for example a GPU) while the extractor container will use only a fraction of the CPU.

The concurrency is solved by using a "max_queue_size" in the pipeline spec, in a similar pattern to how it is solved today in Pachyderm. So the worker won't be utilized in the first datum it will be significantly more efficient for the rest of the job: similar to how a "max_queue_size" allows for a big boost when working from pfs.

**Comment by jdoliner** [comment]
The `max_queue_size` in Pachyderm only works because Pachyderm knows how to separate the files into datums and download one datum at a time, pfs was built specifically to support that. There isn't a general way to do that though, if all we have is a command that downloads the data all we can do is run that command and download the data, all of the data. The extractor code would have to understand how to download a specific chunk of the data, which I think would put too big a burden on

**Comment by RaananHadar** [comment]
I completely agree with you but I guess it is my fault for not being completely clear here: 

I believe that since I am not as intimately familiar with the inner workings of the storage container my analogy may have been inaccurate. Let me explain this from a normal Pachyderm user's eyes, as the use case I am suggesting is very much the same as how Pachyderm is used today which is good, because it doesn't introduce radical change for the user.

Similar to how a transformer code has no notion of the data it transforms (this is supplied by Pachyderm) I would expect the extractor interface to behave **just the same**. Lets say that I've defined in my hypothetical pipeline spec something like this:
```
"extract_input": {
    "pfs": {
      "repo": "image_paths",
      "glob": "/*"
    }
}
```
The image_paths is a normal pfs repo and each input datum will have all the metadata needed to download that specific chunk of data for the transform. So we are still using Pachyderm to do all the datum separation. Its just that there is now a custom extractor that runs concurrently to the transformer.

**Comment by mindthevirt** [comment]
<div>Agent Dale Georg linked Freshdesk ticket <a href="https://pachyderm-support.freshdesk.com/a/tickets/270">270</a> for this issue.</div>
